File: smoketests/tests/add_remove_index.py
from .. import Smoketest, random_string

class AddRemoveIndex(Smoketest):
    AUTOPUBLISH = False

    MODULE_CODE = """
use spacetimedb::{ReducerContext, Table};

#[spacetimedb::table(name = t1)]
pub struct T1 { id: u64 }

#[spacetimedb::table(name = t2)]
pub struct T2 { id: u64 }

#[spacetimedb::reducer(init)]
pub fn init(ctx: &ReducerContext) {
    for id in 0..1_000 {
        ctx.db.t1().insert(T1 { id });
        ctx.db.t2().insert(T2 { id });
    }
}
"""
    MODULE_CODE_INDEXED = """
use spacetimedb::{ReducerContext, Table};

#[spacetimedb::table(name = t1)]
pub struct T1 { #[index(btree)] id: u64 }

#[spacetimedb::table(name = t2)]
pub struct T2 { #[index(btree)] id: u64 }

#[spacetimedb::reducer(init)]
pub fn init(ctx: &ReducerContext) {
    for id in 0..1_000 {
        ctx.db.t1().insert(T1 { id });
        ctx.db.t2().insert(T2 { id });
    }
}

#[spacetimedb::reducer]
pub fn add(ctx: &ReducerContext) {
    let id = 1_001;
    ctx.db.t1().insert(T1 { id });
    ctx.db.t2().insert(T2 { id });
}
"""

    JOIN_QUERY = "select t1.* from t1 join t2 on t1.id = t2.id where t2.id = 1001"

    def between_publishes(self):
        """
        The test `AddRemoveIndexAfterRestart` in `zz_docker.py`
        overwrites this method to restart docker between each publish,
        otherwise reusing this test's code.
        """
        pass

    def test_add_then_remove_index(self):
        """
        First publish without the indices,
        then add the indices, and publish,
        and finally remove the indices, and publish again.
        There should be no errors
        and the unindexed versions should reject subscriptions.
        """

        name = random_string()

        # Publish and attempt a subscribing to a join query.
        # There are no indices, resulting in an unsupported unindexed join.
        self.publish_module(name, clear = False)
        with self.assertRaises(Exception):
            self.subscribe(self.JOIN_QUERY, n = 0)

        self.between_publishes()

        # Publish the indexed version.
        # Now we have indices, so the query should be accepted.
        self.write_module_code(self.MODULE_CODE_INDEXED)
        self.publish_module(name, clear = False)
        sub = self.subscribe(self.JOIN_QUERY, n = 1)
        self.call("add", anon = True)
        self.assertEqual(sub(), [{'t1': {'deletes': [], 'inserts': [{'id': 1001}]}}])

        self.between_publishes()

        # Publish the unindexed version again, removing the index.
        # The initial subscription should be rejected again.
        self.write_module_code(self.MODULE_CODE)
        self.publish_module(name, clear = False)
        with self.assertRaises(Exception):
            self.subscribe(self.JOIN_QUERY, n = 0)

File: crates/smoketests/tests/smoketests/add_remove_index.rs
use spacetimedb_smoketests::Smoketest;

const JOIN_QUERY: &str = "select t_1.* from t_1 join t_2 on t_1.id = t_2.id where t_2.id = 1001";

/// First publish without the indices,
/// then add the indices, and publish,
/// and finally remove the indices, and publish again.
/// There should be no errors
/// and the unindexed versions should reject subscriptions.
#[test]
fn test_add_then_remove_index() {
    let mut test = Smoketest::builder()
        .precompiled_module("add-remove-index")
        .autopublish(false)
        .build();

    // TODO: Does the name do anything? Other tests just let the DB assign.
    let name = format!("test-db-{}", std::process::id());

    // Publish and attempt a subscribing to a join query.
    // There are no indices, resulting in an unsupported unindexed join.
    test.publish_module_named(&name, false).unwrap();
    let result = test.subscribe(&[JOIN_QUERY], 0);
    assert!(result.is_err(), "Expected subscription to fail without indices");

    // Publish the indexed version.
    // Now we have indices, so the query should be accepted.
    test.use_precompiled_module("add-remove-index-indexed");
    test.publish_module_named(&name, false).unwrap();

    // Subscribe and hold across the call, then collect results
    let sub = test.subscribe_background(&[JOIN_QUERY], 1).unwrap();
    test.call_anon("add", &[]).unwrap();
    let results = sub.collect().unwrap();
    assert_eq!(results.len(), 1, "Expected 1 update from subscription");

    // Publish the unindexed version again, removing the index.
    // The initial subscription should be rejected again.
    test.use_precompiled_module("add-remove-index");
    test.publish_module_named(&name, false).unwrap();
    let result = test.subscribe(&[JOIN_QUERY], 0);
    assert!(result.is_err(), "Expected subscription to fail after removing indices");
}
